### DISCUSSION TITLE
Align 2D origin (0, 0) to screen viewport center instead of top-left corner

### DISCUSSION THREAD

**Discussion by geekley** [godotengine/godot-proposals#9719] [2D]
Having the 2D (0, 0) origin align to the viewport/screen center would be much better for game development than the current way that aligns at the top left, specially for 2D games, but also in general.

Please add at least a project setting letting us make all scenes use this logic. I'd even go as far as making it default in new projects (but obviously keeping the option if you need the current behavior).
You could put it at `display/window/world/2d_origin_alignment` = `center` | `top_left`.

### Reasoning
The screen center is always special in both Node2D and Control, all designs will consider it in one way or another.
But screen top-left is NOT particularly special in most usages of Node2D, and in Control you can easily use anchors to deal with the screen corners as you should: independently of resolution and aspect ratio.

Thus an option for screen center origin alignment would greatly facilitate working independently of screen factor. It would specially help in 2D scenes with only central (or no) screen movement, but also in cases like top-down games where you start at (0, 0). I'd go as far as saying even 2D aspects of 3D games would benefit from this.

It also makes much more sense in the editor for "prefab" scenes, i.e. a 2D object to be instantiated. If (0, 0) is at center, the "show viewport" feature is useful to have a size comparison between the object and the screen, as the object will be at the center. Plus you would be able to preview its behavior with <kbd>F6</kbd> and see it in center without having to add a camera or into another scene.

It would be easier to work with Control and Node2D interchangeably.
Most designs are based around the center of the screen (or of parent node), with corners treated specially.
It would also help a lot with any designs that deal with any sort of symmetry.

Lastly there's also the very minor reason of floating-point values being more accurate near the origin, so favoring designs that put world's center around (0, 0) might have this additional little stability benefit. Maybe negligible, but I thought it's worth mentioning it.

**Comment by Sauermann** [comment]
When using `Camera2D`, have you had a look at [`Camera2D.anchor_mode`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-anchor-mode)?

**Reply by geekley** [reply]
Yes, I'm already taking that in consideration. It only affects Node2D stuff.
The issue is stuff you do with Control (like UI, maybe a background) being misaligned with the origin (and your game world) is bad for design. Having this alignment fundamentally in the screen would improve it at lot, specially for those use cases mentioned.
This would align the "show viewport" outline in editor, and also the actual values in-game of "screen px" units, all around (0, 0).

So, basically, you would properly design your UI aligned with the initial game world position.

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot-proposals/issues/1313.

This would be handy for some of the official demos (so they don't need a dummy Camera2D node), but this needs to be done in a way that doesn't break compatibility. This means we can't change how the 2D coordinate system works.

**Reply by geekley** [reply]
I suggested it being a project-wide setting exactly for this reason. So that it's used mostly only in new projects, to avoid breaking stuff in scripts that rely on the assumption that the 2D coordinate system has (0, 0) at top left of the screen.
I believe custom scripts can also benefit from using the logic of: (0, 0) is screen center, everything else varies responsively according to window/screen dimensions.

**Reply by Calinou** [reply]
With the project setting, I wouldn't change the coordinate system. Instead, I'd only change where the fallback Camera2D spawns (when there is no active Camera2D in the current scene).

**Reply by geekley** [reply]
@Calinou Ah but like I said, just having a centered Camera2D on the scene alone would not solve the issue, because that affects just Node2D, not Control. It's not the core of the problem.

What I'm asking for is compatibility between Node2D and Control, so the "show viewport" outline needs to be affected too, centered at (0, 0). Just adding a centered Camera2D to the scene doesn't fix that.

For example, if you need to implement touchscreen controls for moving, you may need to use `TouchScreenButton` (e.g. because of `passby_press`) and that's a Node2D. But you may need to position it like a Control, so you would naturally make it a child of a Control so that you can be properly responsive to any screen dimensions without custom code.
But even if you can make that setup work at runtime, ***it won't appear right in the editor*** because Node2D follows the Camera2D which is centered at origin, but Control follows the viewport which is offset as its origin is at top-left.

So whichever solution is used, it needs to fix the viewport alignment too.

**Reply by geekley** [reply]
Is there a similar "default" CanvasLayer, technically speaking?
I suspect one thing that might help solving it is adding `anchor_mode` to CanvasLayer, like how it works in Camera2D, and making the proposed project setting also apply to the default CanvasLayer, if that makes sense?

**Comment by hsandt** [comment]
For "prefab" / scene objects to instantiate in other scenes, there is this discussion:

Related to Add ability to center editor camera on scenes when running current scene
https://github.com/godotengine/godot-proposals/discussions/8970

Suggestion solutions are more editor-focused than release-focused, though.

**Comment by ryanbeard82** [comment]
Would love to see a solution for this as well. Having my canvas layer elements (HUD, UI, etc) misaligned in the editor is a nightmare and leads to a very inefficient trial-and-error design approach.
<img width="757" alt="Screenshot 2024-09-07 at 10 46 40 AM" src="https://github.com/user-attachments/assets/9e69aae0-bf6d-4a37-9597-fc72cd6be91d">

**Reply by bdrelling** [reply]
This is the exact problem I'm having. Either the editor looks janky, or you have to manually re-center every single one of your scenes, which is absolutely not a good idea.

Folks putting a 👎🏻 on this topic should explain how they avoid this problem.

**Reply by timothyqiu** [reply]
@bdrelling I can see some reasons for not implementing this. In short, not everyone think this is a "problem".

Currently, re-centering the scene is only necessary the first time you open it. This process requires just one mouse drag, but not everyone realizes or cares about it.

However, once the toggle feature is implemented, you will need to manually decide the mode for every scene you create. This can be a burden for those who don't care.

**Reply by geekley** [reply]
I'm not sure you understand the issue.
The problem with "just" re-centering by just applying a specific px offset is that it ties the layout to a specific screen dimension/factor, which is a problem if you want to design independently of screen resolution/aspect.

If you could set a project setting to have all scenes use 2D center at (0, 0) for everything (both Node2D and Control), then you can design independently of screen resolution. Which is a must for an engine where one of the main selling points is being multi-platform.

**Reply by bdrelling** [reply]
> Currently, re-centering the scene is only necessary the first time you open it. This process requires just one mouse drag, but not everyone realizes or cares about it.

@timothyqiu Can you elaborate? I'm not following. When you're talking about "one mouse drag", what are you dragging and in what context to fix this? Are you saying that it is possible to move the default viewport already outside of requiring project settings? If so, how?

Above, I'm specifically referring to the problem above of having UI fixed to the scene viewport, such that it dictates where the center of all scenes would need to be for it to not look strange when showing up in the scene editor. @ryanbeard82's example is precisely what I mean.

What is the solution for getting around that?

I hope my tone doesn't sound confrontational, I really don't understand and don't have enough knowledge of Godot to know what else would prevent this.

**Reply by timothyqiu** [reply]
I have no problem with this feature. I'm just trying to explain that it's perfectly normal for not everyone to see this as a significant problem, thus thumb-down reactions.

**Reply by bdrelling** [reply]
@timothyqiu I think there's a misunderstanding: I am not referring to [the feature request](https://github.com/godotengine/godot-proposals/discussions/9719#discussion-6638619) at all, I am referring to [the exact problem ryanbeard82 is describing above](https://github.com/godotengine/godot-proposals/discussions/9719#discussioncomment-10577756).

I understand why people might downvote the topic, there's no confusion there. What I'm saying in [my comment](https://github.com/godotengine/godot-proposals/discussions/9719#discussioncomment-11022603) is that people who are downvoting should _also_ explain how they are avoiding the [UI alignment issue](https://github.com/godotengine/godot-proposals/discussions/9719#discussioncomment-10577756) as it would help myself and others to learn, and to see a counterpoint to the proposed solution. Downvoting without context isn't typically helpful.

So [you mentioned above](https://github.com/godotengine/godot-proposals/discussions/9719#discussioncomment-11022903) that there's some process requiring one mouse drag or some solution you had to avoiding the problem. I'm asking you to elaborate because I don't understand and it might help myself and others.

Could you elaborate further, please?

**Comment by timothyqiu** [comment]
This could be a per-scene metadata. And it would also help the editor generate better scene thumbnails.

**Reply by geekley** [reply]
I agree that a per-scene *override* option would be an improvement. But only if I can still have a project-wide option that I could set to `center` by default. After all, usually most "scenes" in a project are actually "prefabs" meant to be instantiated (so center is better), and the ones actually representing top-level game "screens" are the minority (where one might want top-left for some reason).

**Comment by MrJake222** [comment]
Hey,

I'd like to see it implemented too. Up until now all my scenes are centered around `(0, 0)`. And I want a root scene to be also `(0,0)` centered, for the UI placement on different `CanvasLayer`.

**Comment by hubertgrzeskowiak** [comment]
I just started learning this engine and the chaos around coordinate system is a real deal breaker for me. It's a small issue, but such a fundamental design flaw.

New scenes start with 0,0 in the top left corner, and that's also how you see them when previewing individual scenes. So if you wanted an object to be in the center of the visible scene in both editor and when previewing, you need to position the objects at half the screen size in x and y. But there is no constants, no buttons, nothing to indicate that or to help with that.

It just doesn't make sense to me...

**Comment by whitemagehealing** [comment]
Hi! I think I found the solution by hunting around the Game Tab interface.

1. Enable the "Override the in-game camera."
2. Select "Manipulate From Editors".

I have attached some photos so you can see what I mean. This should autoshift your Scene to be fully viewable without having to destructively edit parameters. It also looks like it works PER SCENE, so it's not a global effect for everything. This is great, because there are Scenes which we DON'T want to move, like full play screens we want to see the "compiled" (end user view) version of.

**EDIT: Actually, nevermind, that's not what you are talking about. However, I'm just going to leave my comment here for others that might accidentally think the original problem is what I thought it was. If the devs feel it's not helpful, they are free to delete it.**

![image](https://github.com/user-attachments/assets/56823335-f899-4476-8f26-3244c732aa5e)
![image](https://github.com/user-attachments/assets/9dc70774-0de9-4f18-8e65-3ecd452afddc)

**Comment by MidnatheBlackRobe** [comment]
It shouldn't be that you change the origin, it should instead be that you have an option to center the default viewport and control elements on the origin. The origin of a game should be absolute, (0, 0). It's kind of the definition of origin. However, **it makes practically no sense that the default camera and ui elements aren't aligned with the origin, *and* you have no way to change it.**

Here are some common arguments against this and why I believe they're wrong:

Argument: This is how it is in all previous projects, so it would break things.
Counter: That's why it should be an opt-in option. Don't get me wrong, I'm against this line of reasoning in general and I think it prevents change, but if you didn't want to make it the default, at least give people the ability to change it.

Argument: It reflects the low-level architecture of how the screen coordinate system works.
Counter: Okay....... so? It's inconsistent with how Node2Ds are centered in Godot, and Godot abstracts a ton of low-level architecture to make projects easier to make and work with.

Argument: It would take additional processing power
Counter: You're not wrong, but this would be such a tiny optimization that it obviously isn't worth the massive headache that is things not lining up properly by default. Readability trumps trivial optimization every time.

Argument: It wouldn't change properly with different screen sizes
Counter: We literally have a whole anchoring system already that specifies what part of the screen it's anchored to. Not to mention, I'd argue that the center of the screen is the most important part to preserve for most games anyways.

It just doesn't make any sense to me why the people developing and contributing to Godot haven't changed this, or why they wouldn't have made this the default from the start.
